### PR TITLE
add 1f1b and fix interleaved 1f1b

### PR DESCRIPTION
**Pull request by H-Huang**
add non-interleaved 1f1b and fix interleaved 1f1b

- interleaved 1f1b (looped dfs) was hanging since the backward stage was not being retrieved correctly. Removed a lot of code and simplified it.
- added some tests (TODO will add more)

### CODE REVIEW

File: pippy/PipelineSchedule.py [hunk 1]
```diff
@@ -592,39 +674,23 @@ def step(self, microbatches):
...
-
-        self.stage_id_to_global_stage_id = [
-            (i * n_pp) + pp_id for i in range(self.n_local_stages)
+        stage = stages[0]
+        self.world_size = stage.world_size
+        self.rank = stage.rank
+        self.total_stages = self.n_local_stages * self.world_size
+        self.stage_idx_to_global_stage_id = [
+            (i * self.world_size) + self.rank
+            for i in range(self.n_local_stages)
```
Comment (kwen2501) [inline pippy/PipelineSchedule.py:691]: better way is:
```
self.local_idx_to_global_stage_id = [
    stage.id for stage in self.stages
]
```

File: pippy/PipelineSchedule.py [hunk 2]
```diff
@@ -645,155 +711,112 @@ def step(self, microbatches):
         Rank 1:    1F 1F 1F 1F 3F3B 3F 3F 3F
         """
 
+        n_microbatch = len(microbatches)
+        assert (
+            n_microbatch % self.world_size == 0
+        ), f"Looped DFS schedule requires microbatch_size ({n_microbatch}) to be a multiple of the number of \
+pipelined ranks ({self.world_size})"
```
Comment (kwen2501) [inline pippy/PipelineSchedule.py:718]: aha, shall we rename `world_size` as `pp_group_size`?

File: pippy/PipelineSchedule.py [hunk 3]
```diff
@@ -645,155 +711,112 @@ def step(self, microbatches):
...
-            f"rank {self.pp_id} - minibatch_index {[minibatch_index(step) for step in range(self.total_steps)]}"
-        )
-        logger.info(
-            f"rank {self.pp_id} - stage_index {[stage_index(step) for step in range(self.total_steps)]}"
-        )
+        for step in range(self.total_steps):
+            # warmup, forward only
+            if step < warmup_steps:
+                fwd_stage = self.stages[forward_stage_index(step)]
+                mb_index = minibatch_index(step)
```
Comment (kwen2501) [inline pippy/PipelineSchedule.py:768]: `microbatch` instead of `minibatch`?

File: pippy/PipelineSchedule.py [hunk 4]
```diff
@@ -645,155 +711,112 @@ def step(self, microbatches):
...
-                step / (self.seq_size * self.n_local_stages)
+            return (step % self.world_size) + self.world_size * int(
+                step / (self.world_size * self.n_local_stages)
             )
 
-        def stage_index(step):
-            # step: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15
-            # index:0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1,  1,  0,  0,  1,  1
-            return int((step / self.seq_size) % self.n_local_stages)
+        def forward_stage_index(step):
```
Comment (kwen2501) [inline pippy/PipelineSchedule.py:753]: Do you mean local stage index here?

**Review by kwen2501 — COMMENTED** [review]
Both schedules have structures like:
```
for step in range(total_steps):
    if step < ... :
        ...
    elif ... <= step < ...:
        ...
    else:
        ...
```
I think it would be more readable if written as:
```
for step in range(0, warm_up):
    ...
for step in range(warm_up, warm_up + 1f1b):
    ...
for step in range(warm_up + 1f1b, total):
    ...
```

File: pippy/PipelineSchedule.py [hunk 5]
```diff
@@ -559,6 +569,78 @@ def step(self, microbatches):
...
+
+                    self._stage.forward(microbatches[i])
+
+                    ops = self._stage.get_fwd_send_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops)
+            if (
+                warmup_steps
+                <= i
+                < warmup_steps + main_1f1b_steps + cooldown_steps
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:629]: thats true, i dont need the right side inequality

File: pippy/PipelineSchedule.py [hunk 6]
```diff
@@ -559,6 +569,78 @@ def step(self, microbatches):
...
+        total_ops = num_microbatches * 2
+
+        # Example, 4 GPUs, 8 microbatches
+        # Stage 0: 6 warmup, 2 1f1b, 6 cooldown
+        # Stage 1: 4 warmup, 4 1f1b, 4 cooldown
+        # Stage 2: 2 warmup, 6 1f1b, 2 cooldown
+        # Stage 3: 0 warmup, 8 1f1b, 0 cooldown
+        # fwd only
+        warmup_steps = min(
+            num_microbatches, 2 * (self.world_size - self.stage_id - 1)
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:592]: changed to `pp_group_size` based on your other comment. Thanks!

File: pippy/PipelineSchedule.py (hunk 2 above)
Comment (H-Huang) [reply pippy/PipelineSchedule.py:718]: Updated, thanks!

File: pippy/PipelineSchedule.py (hunk 4 above)
Comment (H-Huang) [reply pippy/PipelineSchedule.py:750]: yes the local index, will rename

File: pippy/PipelineSchedule.py [hunk 7]
```diff
@@ -559,6 +569,78 @@ def step(self, microbatches):
...
+                with record_function(f"Backward {i}"):
+                    ops = self._stage.get_bwd_recv_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops).pop().wait()
+
+                    self._stage.backward()
+
+                    ops = self._stage.get_bwd_send_ops()
+                    if ops:
+                        dist.batch_isend_irecv(ops)
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/PipelineSchedule.py:637]: `warmup_steps` and `main_1f1b_steps` always include `forward()`, and `main_1f1b_steps` and `cooldown_steps` always include `backward()` so this logic combines that

**Review by kwen2501 — APPROVED** [review]
LGTM. Thanks for the attending to the comments :)

**Pull request merged**